src/openrct2/rct2/SeaDecrypt.cpp

```diff
@@ -0,0 +1,109 @@
+/*****************************************************************************
+ * Copyright (c) 2014-2020 OpenRCT2 developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/OpenRCT2/OpenRCT2
+ *
+ * OpenRCT2 is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#include "../common.h"
+#include "../core/File.h"
+#include "../core/Path.hpp"
+#include "RCT2.h"
+
+#include <cstdint>
+#include <memory>
+#include <string_view>
+
+constexpr int32_t MASK_SIZE = 0x1000;
+
+struct EncryptionKey
+{
+    uint32_t Seed0{};
+    uint32_t Seed1{};
+};
+
+static EncryptionKey GetEncryptionKey(const std::string_view& fileName)
+{
+    auto fileNameLen = static_cast<int32_t>(fileName.size());
+    uint32_t s0 = 0;
+    for (int i = fileNameLen - 1; i >= 0; i--)
+    {
+        s0 = (s0 + (s0 << 5)) ^ fileName[i];
+    }
+
+    uint32_t s1 = 0;
+    for (int i = 0; i < fileNameLen; i++)
+    {
+        s1 = (s1 + (s1 << 5)) ^ fileName[i];
+    }
+
+    return EncryptionKey{ s0, s1 };
+}
+
+static std::unique_ptr<uint8_t[]> CreateMask(const EncryptionKey& key)
+{
+    auto result = std::make_unique<uint8_t[]>(MASK_SIZE);
```

Comment:
A std vector would work fine for this and is generally the preferred default unless absolute speed is required.